core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -340,23 +340,48 @@ class GroupCoordinator(val brokerId: Int,
 
   def handleDeleteGroups(groupIds: Set[String]): Map[String, Errors] = {
     if (!isActive.get) {
-      groupIds.map((_ -> Errors.COORDINATOR_NOT_AVAILABLE)).toMap
+      groupIds.map(_ -> Errors.COORDINATOR_NOT_AVAILABLE).toMap
     } else {
       var groupErrors: Map[String, Errors] = Map()
-      var eligibleGroups: Seq[String] = Seq()
+      var eligibleGroups: Seq[GroupMetadata] = Seq()
 
       groupIds.foreach { groupId =>
         if (!validGroupId(groupId))
-          groupErrors += (groupId -> Errors.INVALID_GROUP_ID)
-        else if (!isCoordinatorForGroup(groupId)) {
-          groupErrors += (groupId -> Errors.NOT_COORDINATOR)
-        } else if (isCoordinatorLoadInProgress(groupId))
-          groupErrors += (groupId -> Errors.COORDINATOR_LOAD_IN_PROGRESS)
-        else
-          eligibleGroups ++= Seq(groupId)
+          groupErrors += groupId -> Errors.INVALID_GROUP_ID
+        else if (!isCoordinatorForGroup(groupId))
+          groupErrors += groupId -> Errors.NOT_COORDINATOR
+        else if (isCoordinatorLoadInProgress(groupId))
+          groupErrors += groupId -> Errors.COORDINATOR_LOAD_IN_PROGRESS
+        else {
+          groupManager.getGroup(groupId) match {
+            case None =>
+              groupErrors += groupId -> Errors.GROUP_ID_NOT_FOUND
```

Comment:
This case should be handled the same as if the group is dead. You can probably add a little helper to avoid the duplication.

Reply:
On second thought, maybe that's overkill. I'm trying to think how we can be sure that we're returning this error code correctly. Maybe we need to check for group existence while holding the `ownedPartitions` lock in `GroupMetadataManager`.

Reply:
If `ownedPartitions` includes the corresponding topic partition for the group, and if the cached group either doesn't exist or is Dead, then I think it is safe to return GROUP_ID_NOT_FOUND. Maybe we can just add a method like the following to `GroupMetadataManager`:

```scala
// return true iff group is owned and the group doesn't exist
def groupNotExists(groupId: String) = inLock(partitionLock) {
  isGroupLocal(groupId) && (!groupMetadataCache.contains(groupId) || groupMetadataCache.get(groupId).is(Dead))
}
```

Then we can use this function here instead of checking the coordinator again. The name could probably be improved.

Reply:
For both `case None` and `case Dead` we already know the second part (`(!groupMetadataCache.contains(groupId) || groupMetadataCache.get(groupId).is(Dead))`) is true. So, it suffices to check `isGroupLocal(groupId)` (to avoid redundant checks). Is that correct? If so, we wouldn't need this helper (at least here).

Reply:
The point is to check it while holding the partition lock so that it is an atomic operation. This ensures that we will not have any race conditions with partition loading/unloading.

Reply:
Aah, right. That makes sense.